src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1793,93 +1865,166 @@ private boolean unsignedLongCompare(long one, long two) {
+        if ((x.intLen & 1) != 0)
+            shift += 32; // x.intLen must be even
+
+        if (shift != 0) {
+            x = new MutableBigInteger(x);
+            x.leftShift(shift);
+        }
+
+        // Compute sqrt and remainder
+        MutableBigInteger[] sqrtRem = x.sqrtRemZimmermann(x.intLen, needRemainder);
+
+        // Unnormalize
+        if (shift != 0) {
+            final int halfShift = shift >> 1;
+            if (needRemainder) {
+                // shift <= 62, so s0 is at most 31 bit long
+                final long s0 = sqrtRem[0].value[sqrtRem[0].offset + sqrtRem[0].intLen - 1]
+                        & (-1 >>> -halfShift); // Remove excess bits
+                if (s0 != 0L) { // An optimization
+                    MutableBigInteger doubleProd = new MutableBigInteger();
+                    sqrtRem[0].mul((int) (s0 << 1), doubleProd);
+
+                    sqrtRem[1].add(doubleProd);
+                    sqrtRem[1].subtract(new MutableBigInteger(s0 * s0));
                 }
+                sqrtRem[1].rightShift(shift);
+            }
+            sqrtRem[0].primitiveRightShift(halfShift);
+        }
+        return sqrtRem;
+    }
 
-                xk = xk1;
-            } while (true);
-        } else {
-            // Set up the initial estimate of the iteration.
+    private static long ulongSqrt(long x) {
+        /* For every long value s in [0, 2^32) such that x == s * s,
+         * it is true that s - 1 <= (long) Math.sqrt(x >= 0 ? x : x + 0x1p64) <= s,
+         * and if x == 2^64 - 1, then (long) Math.sqrt(x >= 0 ? x : x + 0x1p64) == 2^32.
+         * Since both cast to long and `Math.sqrt()` are (weakly) increasing,
+         * this means that the value returned by Math.sqrt()
+         * for a long value in the range [0, 2^64) is either correct,
+         * or rounded up/down by one if the value is too high
+         * and too close to a perfect square.
+         */
+        long s = (long) Math.sqrt(x >= 0 ? x : x + 0x1p64);
+        long s2 = s * s;  // overflows iff s == 2^32
+        return Long.compareUnsigned(x, s2) < 0 || s > LONG_MASK
+                ? s - 1
+                : (Long.compareUnsigned(x, s2 + (s << 1)) <= 0 // x <= (s + 1)^2 - 1, does not overflow
+                        ? s
+                        : s + 1);
+    }
 
-            // Obtain the bitLength > 63.
-            int bitLength = (int) this.bitLength();
-            if (bitLength != this.bitLength()) {
-                throw new ArithmeticException("bitLength() integer overflow");
-            }
+    /**
+     * Assume {@code 2 <= len <= intLen && len % 2 == 0
+     * && Integer.numberOfLeadingZeros(value[offset]) <= 1}
+     * @implNote The implementation is based on Zimmermann's works available
+     * <a href="https://inria.hal.science/inria-00072854/en/">  here</a> and
+     * <a href="https://www.researchgate.net/publication/220532560_A_proof_of_GMP_square_root">  here</a>
```

Comment:
The following variant should be preferred, as it has a readable figure on p. 21, whereas the same figure in the variant linked in the PR seems broken for some reason.
https://inria.hal.science/inria-00072113/document

Reply:
I'm referring to the figure that depicts the memory layout during the core algorithm (p.22 resp. p.21, depending on the variant).
I tried on macOS with the standard Preview app, on Windows with Adobe Reader, on kubuntu with the standard reader, and with the reader built in Firefox.

Reply:
Now I noticed. I went to the wrong page.